### DISCUSSION TITLE
Issue with exception_handler not catching exception

### DISCUSSION THREAD

**Discussion by dhcsousa** [fastapi/fastapi#11471] [Questions] [answered]
# Introduction

When utilizing asynchronous endpoints with FastAPI and attempting to handle custom exceptions using the exception_handler, it seems that the exception_handler does not catch custom exceptions raised within asynchronous generators

### Expected Behavior

The exception_handler should catch and handle custom exceptions raised within asynchronous generators, returning the specified response.

### Actual Behavior
The exception_handler fails to catch the custom exceptions raised within asynchronous generators, leading to unhandled exceptions and incorrect responses being returned to the client.

### Example Code

```python
from collections.abc import AsyncGenerator

from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse
from fastapi.testclient import TestClient
from sse_starlette import EventSourceResponse
from starlette.middleware.exceptions import ExceptionMiddleware

app = FastAPI()


# Custom exception class
class CustomException(Exception):
    def __init__(self, code: int, message: str):
        self.code = code
        self.message = message


@app.get("/test")
def test():
    raise CustomException(message="Example", code=400)


async def some_async_stream() -> AsyncGenerator[str, None]:
    # Simulate some asynchronous stream
    yield "item1"
    yield "raise_exception"
    yield "item2"


# Define an asynchronous generator that yields elements from the stream
async def stream_generator() -> AsyncGenerator[str, None]:
    # Assume you have some asynchronous stream of elements
    async for item in some_async_stream():
        yield item


# Define the test_async endpoint
@app.get("/test_async")
async def test_async(request: Request):
    async def custom_stream_response() -> AsyncGenerator[str, None]:
        async for item in stream_generator():
            if item == "raise_exception":
                raise CustomException(message="Example", code=400)
            yield item

    # Create a streaming response using the custom_stream_response function
    return EventSourceResponse(custom_stream_response())


@app.exception_handler(CustomException)
async def openapi_generic_exception_handler(
    request: Request, exc: CustomException
) -> JSONResponse:
    return JSONResponse(
        status_code=400, content="Should return this"
    )  # Internal Server Error


app.add_middleware(ExceptionMiddleware, handlers=app.exception_handlers)

tc = TestClient(app)

response = tc.get("/test")
assert (
    response.json() == "Should return this"
)  # this assertion passes, exception_handler does not catch CustomException

response = tc.get(
    "/test_async"
)  # This immediately fails, exception_handler does not catch CustomException
```

### Operating System

macOS

### FastAPI Version

0.100.1

### Pydantic Version

2.6.4

### Python Version

3.11.8

**Comment by YuriiMotov** [comment, marked as answer]
Logically speaking, it's useless to handle exceptions and return `JSONResponse(status_code=400)` after FastAPI already started sending response (I'm talking about `/test_async` endpoint).
Response headers had been already sent at that moment and it's impossible to change them.

So, using `exception_handler` you can only handle exceptions that were raised before the endpoint function's execution finished (before you return `EventSourceResponse(custom_stream_response())`. At this moment FastAPI sends response headers and starts iterating through the stream response generator function to send response content. And if any exception happens, it leads to server error 500.

**Reply by dhcsousa** [reply]
Thank you for your response.

In that case, how would you handle a situation where an exception is raised during a streaming event? Suppose you have a scenario where FastAPI should stream data as it arrives without buffering.

**Reply by YuriiMotov** [reply]
You can wrap the code of `stream_generator` by `try .. except` and handle exceptions on server side.
You can either send the error info to the client as a chunk of stream response (if your client can handle it) or you can use [callback mechanism](https://fastapi.tiangolo.com/advanced/openapi-callbacks/) to notify client about errors during streaming.
